### DISCUSSION TITLE
Existing Dakota 6.16 install broken after upgrade to Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by sitadrost** [snl-dakota/dakota#25] [Q&A] [answered]
On Ubuntu 20.04, I used to have a working Dakota 6.16.0, installed from source. However, after upgrading to Ubuntu 22.04, Dakota no longer works, because it turns out that, in the process of upgrading, Boost was removed for some reason.

In the meantime, I've re-installed Boost, but Dakota can't find it:
`
dakota: error while loading shared libraries: libboost_filesystem.so.1.71.0: cannot open shared object file: No such file or directory
`
When I did a search for this particular file, I found it hidden in some Snap directory, so I can't really blame Dakota for not being able to find it:
`/snap/kde-frameworks-5-91-qt-5-15-3-core20/1/usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.71.0`

How do I tell Dakota where to look? Do I need to re-make/re-build? And has anyone else on Ubuntu 22.04 experienced this type of issue?

Also, I'm not sure of this Snap thing; according to Ubuntu, "Snaps are several things, all at once. They are confined, standalone Linux applications that bundle all their necessary dependencies, which means they do not need to rely on the underlying system, and can run independently of it." I don't know, but Boost as a confined, standalone application doesn't seem to make much sense to me, so perhaps I should be looking for a non-Snap version instead?

EDIT: 
Ah, wait, sorry: apparently this Snap-directory contains an old version. Turns out I also have a newer version (1.74.0) in a more straightforward location: `/usr/lib/x86_64-linux-gnu/`

So, ignoring this Snap-directory, how do I tell Dakota to use this newer version of Boost, and where to find it?

**Comment by jadamstephens** [comment]
HI,

You can run cmake with the argument `-D BOOST_ROOT:PATH=/usr/lib/x86_64-linux-gnu/`. Equivalently, if you have created a CMake cache file that you are loading using `-C` when you run cmake, you can add a variable setting:

``` cmake
set(BOOST_ROOT "/usr/lib/x86_64-linux-gnu/" CACHE PATH "Boost root directory")
```
This assumes that the aforementioned directory really does contain include and lib folders that, in turn, include headers and libraries for Boost.

You may need to install a Boost "devel" package using apt in order to get things to work. A quick google search suggests that the name of the package you should install is `libboost-all-dev`. (Many Linux distros provide two flavors of packages for libraries such as Boost. One package type installs just what is needed to satisfy dependencies of other packages, and another, the "devel" package type, provides extra things that are needed to link to the library when building from source.)

I'm not completely certain 1.74 will work; it may be too new. We haven't done a good job of investigating our maximum allowed versions of many of Dakota's dependencies. If you get Boost-related build errors, that's a likely cause.

**Comment by sitadrost** [comment]
Thanks, I'll give that a try. The package that I (re-)installed is `libboost-all-dev`, so that should be alright. I'll let you know whether version 1.74.0 works

**Reply by sitadrost** [reply]
Oh dear, I'm really sorry, but I lost the Dakota-users mail list e-mail with the build instructions that I used last time...
Running cmake as you suggest, with the `BOOST_ROOT` argument, results in errors, as does following the [instructions for Ubuntu 20.04](https://dakota.sandia.gov/content/linux-ubuntu-2004)
The first error that I encounter is
`
CMake Error at src/CMakeLists.txt:1148 (export):
  export given target "teuchosremainder" which is not built by this project.
`
For sake of completeness, I also attached the full output of cmake, and `CMakeOutput.log` and `CMakeError.log`
Any idea how I should resolve these errors?

Many thanks!
[cmake.log](https://github.com/snl-dakota/dakota/files/10730504/cmake.log)
[CMakeError.log](https://github.com/snl-dakota/dakota/files/10730505/CMakeError.log)
[CMakeOutput.log](https://github.com/snl-dakota/dakota/files/10730506/CMakeOutput.log)

**Reply by jadamstephens** [reply]
It looks like cmake is finding and trying to use some libraries in your old installation of Dakota instead of rebuilding them. You may have Dakota's bin and lib folder on your PATH and LD_LIBRARY_PATH (perhaps you're setting them in your .bashrc or .bash_profile). Try removing those settings, restarting your terminal, and running cmake again.

**Reply by sitadrost** [reply]
Thanks, yes, you're right, Dakota was still in my path, so I removed it:
`
echo $PATH
/usr/local/texlive/2021/bin/x86_64-linux:/home/sita/miniconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
`

Unfortunately, I'm still getting errors, including the one that I mentioned in my previous post, about `"teuchosremainder"` (see also attached log files).

Would it help to start from scratch (and upgrade to 6.17 while I'm at it)? And what would I have to do/remove to do so?
[230220_cmake.log](https://github.com/snl-dakota/dakota/files/10781799/230220_cmake.log)
[CMakeError.log](https://github.com/snl-dakota/dakota/files/10781801/CMakeError.log)
[CMakeOutput.log](https://github.com/snl-dakota/dakota/files/10781802/CMakeOutput.log)

**Reply by briadam** [reply, marked as answer]
Issues with imported CMake targets for Trilinos are often due to having an existing Dakota installation on the same CMAKE_PREFIX_PATH you are currently building for. You can either specify a different CMAKE_PREFIX_PATH, e.g., /home/user/local/dakota2 when running CMake for Dakota, or just remove the existing Dakota installation if it's easy to do so. There is also a workaround where you could try setting DAKOTA_NO_FIND_TRILINOS:BOOL=TRUE, but that is only suggested as a last resort.

**Reply by sitadrost** [reply]
Thanks, great, that worked: I removed my entire existing Dakota installation and started from scratch. Looks like I've now successfully built Dakota 6.17.
Thanks for all your help and patience!
